File: src/Calculator-Lib/EnumerableCollectionNode.cpp
#include "EnumerableCollectionNode.h"

/// <summary>
/// Adds a child and sets its parent
/// </summary>
/// <param name="node">New child node</param>
void EnumerableCollectionNode::AddChild(unique_ptr<ExpNode> node) {
	children_.push_back(move(node));
}

/// <summary>
/// Inserts child between this and its last child
/// </summary>
/// <param name="node">this's new child node</param>
void EnumerableCollectionNode::InsertChild(unique_ptr<BranchNode> node) {
	// Gets this's last child and remove from children_
	node->AddChild(move(children_.back()));
	children_.pop_back();
	AddChild(move(node));
}

/// <summary>
/// Gets child at index
/// </summary>
/// <returns>child at index</returns>
const ExpNode& EnumerableCollectionNode::GetChild(int index) {
	return *children_[index];
}

/// <summary>
/// Gets amount of children
/// </summary>
/// <returns>children count</returns>
int EnumerableCollectionNode::ChildCount() const {
	return children_.size();
}

/// <summary>
/// Gets the priority based on the oper
/// </summary>
/// <returns>Proirity value</returns>
Priority EnumerableCollectionNode::GetPriority() const {
	return Priority::VALUE;
}

File: src/Library/Calculator/ExpressionTree/Nodes/Collections/EnumerableCollectionNode.cs
using Calculator.ExpressionTree.Nodes.Values;
using System.Collections.Generic;

namespace Calculator.ExpressionTree.Nodes.Collections
{
    public abstract class EnumerableCollectionNode : BranchNode
    {
        public EnumerableCollectionNode()
        {
            Children = new List<ExpNode>();
        }

        public EnumerableCollectionNode(EnumerableCollectionNode node)
        {
            Children = new List<ExpNode>();
            foreach (var child in node.Children)
            {
                Children.Add(child.Clone());
            }
        }

        protected List<ExpNode> Children { get; }

        public override int ChildCount => Children.Count;

        public override Priority Priority => Priority.OVERRIDE;

        public override void AddChild(ExpNode node)
        {
            Children.Add(node);
        }

        public override void InsertChild(BranchNode node)
        {
            ExpNode lastChild = Children[ChildCount - 1];
            Children.RemoveAt(ChildCount - 1);
            node.AddChild(lastChild);
            AddChild(node);
        }

        public override void ReplaceChild(ExpNode node, int index)
        {
            Children[index] = node;
        }

        public override ExpNode GetChild(int index)
        {
            return Children[index];
        }

        public override void ClearChildren()
        {
            Children.Clear();
        }

        public override bool IsConstant()
        {
            foreach (var child in Children)
            {
                if (!child.IsConstant()) return false;
            }
            return true;
        }

        public override bool IsConstantBy(VarValueNode variable)
        {
            foreach (var child in Children)
            {
                if (!child.IsConstantBy(variable)) return false;
            }
            return true;
        }
    }
}
